include/rdma/fi_domain.h

```diff
@@ -128,6 +128,7 @@ enum fi_hmem_iface {
 	FI_HMEM_ZE,
 	FI_HMEM_NEURON,
 	FI_HMEM_SYNAPSEAI,
+	FI_HMEM_XPMEM,
```

Comment:
Hmm... does xpmem count as an hmem interface?  I thought it was a different way to setup shared memory transfers.

The approach here looks clean, and this may be an easy way to add support.

Reply:
I was kinda struggling with the same question. The way I convinced myself is "FI_HMEM_SYSTEM", which is the default if you don't have HMEM. XPMEM also needs to use the the IPC Caching (which I was pondering creating a patch to just rename it to caching and remove the IPC, since it can be used for other purposes as well). Also this allows other providers to use it. For example, GNI has its own XPMEM implementation. So if GNI doesn't endup being removed in 2.0, it can theoretically use this infra. Over all this approach seems like it has precedence. It also fits nicely in the entire architecture and the changes in SHM provider to add support is fairly straight forward, same path as CMA.

Reply:
@aingerson - The changes look decent.  Any concerns about this approach?  Could a similar approach be used for CMA or DSA?

Reply:
I agree, the changes themselves look decent. The thing I struggle with if we put it in the HMEM category is that it blurs the lines for how we define HMEM. The target for defining these has been both to indicate the API to use as well as the location of the memory. We've also defined FI_MR_HMEM as needing to register any of the interfaces in this enum. So even though XPMEM matches the op structure nicely, I'm not sure it's the best grouping. At this point, now we're looking at 3 different methods just for copying FI_HMEM_SYSTEM memory (CMA, DSA, XPMEM) so I think we could consider defining a separate (internal) method of outlining these different methods together that doesn't confuse the HMEM interfaces.

Reply:
This is a good point.  We're modifying the user-visible API because it's convenient for the internal implementation.  But is XPMEM something the app should be dealing with directly?

Reply:
So the idea here is to have a shared memory level callback grouping similar to HMEM, and add XPMEM there?
I do think though that this shouldn't be specific to shared memory but make the infrastructure available to other providers for use. Thoughts on that?

Reply:
Yes, agreed. I think we can make it internal and not expose it to the user API but add it as an ofi_<>.h header so anyone can use it. That way gni can use them for its xpmem implementation and any other provider that wants to take advantage of xpmem or DSA can do so as well.

Reply:
sounds good. I'll update the patch set